pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/relocation/RealtimeSegmentRelocationManager.java

```diff
@@ -0,0 +1,287 @@
+    }
+    List<String> completedServers = _helixAdmin.getInstancesInClusterWithTag(_helixManager.getClusterName(),
+        realtimeTagConfig.getCompletedServerTag());
+    if (completedServers.isEmpty()) {
+      throw new IllegalStateException(
+          "Found no realtime completed servers with tag " + realtimeTagConfig.getCompletedServerTag());
+    }
+
+    Map<String, Map<String, String>> mapFields = idealState.getRecord().getMapFields();
+
+    // create map of completed server name to num segments it holds.
+    // This will help us decide which completed server to choose for replacing a consuming server
+    Map<String, Integer> completedServerToNumSegments = new HashMap<>(completedServers.size());
+    for (String server : completedServers) {
+      completedServerToNumSegments.put(server, 0);
+    }
+    for (Map.Entry<String, Map<String, String>> entry : mapFields.entrySet()) {
+      for (String instance : entry.getValue().keySet()) {
+        if (completedServers.contains(instance)) {
+          completedServerToNumSegments.put(instance, completedServerToNumSegments.get(instance) + 1);
+        }
+      }
+    }
+    MinMaxPriorityQueue<Map.Entry<String, Integer>> completedServersQueue =
+        MinMaxPriorityQueue.orderedBy(new Comparator<Map.Entry<String, Integer>>() {
+          @Override
+          public int compare(Map.Entry<String, Integer> entry1, Map.Entry<String, Integer> entry2) {
+            return Integer.compare(entry1.getValue(), entry2.getValue());
+          }
+        }).maximumSize(completedServers.size()).create();
+    completedServersQueue.addAll(completedServerToNumSegments.entrySet());
+
+    // get new mapping for segments that need relocation
+    segmentNameToInstancesMap = createNewSegmentToInstanceStateMap(mapFields, consumingServers, completedServersQueue);
+
+    return segmentNameToInstancesMap;
+  }
+
+  /**
+   * Given an ideal state mapping, list of consuming serves and completed servers,
+   * create a mapping for those segments that should relocate a replica from consuming to completed server
+   * @param mapFields
+   * @param consumingServers
+   * @param completedServersQueue
+   * @return
+   */
+  private Map<String, Map<String, String>> createNewSegmentToInstanceStateMap(
+      Map<String, Map<String, String>> mapFields, List<String> consumingServers,
+      MinMaxPriorityQueue<Map.Entry<String, Integer>> completedServersQueue) {
+
+    Map<String, Map<String, String>> segmentNameToInstancesMap = new HashMap<>();
+    // TODO: we are scanning the entire segments list every time. This is unnecessary because only the latest segments will need relocation
+    // Can we do something to avoid this?
+    // 1. Time boundary: scan only last day whereas runFrequency = hourly
+    // 2. For each partition, scan in descending order, and stop when the first segment not needing relocation is found
```

Comment:
(2) is good, but then we will need to sort the segments anyway. If we have a table that has a long retention for realtime, there may be a lot of segments in there....